src/ocean/util/container/queue/DynamicQueue.d

```diff
@@ -0,0 +1,244 @@
+            full.
+
+    ***************************************************************************/
+
+    override public T* push ( )
+    {
+        if (this.auto_shrink)
+        {
+            if (oldest > this.buffer.length / 3)
+                this.shrink();
+        }
+
+        this.buffer.length = this.buffer.length + 1;
+        return &this.buffer[][$-1];
+    }
+
+    /***************************************************************************
+
+        Pushes an element into the queue.
+
+        Params:
+            element = element to push (will be left unchanged)
+
+        Returns:
+            true on success or false if the queue is full.
+
+    ***************************************************************************/
+
+    override public bool push ( T element )
+    {
+        *this.push() = element;
+        return true;
+    }
+
+    /***************************************************************************
+
+        Pops an element from the queue and returns a pointer to that element.
+        The value of the element must then be copied from the location pointed
+        to before calling push() or pop() the next time.
+
+        Returns:
+            pointer to the element popped from the queue or null if the queue is
+            empty.
+
+    ***************************************************************************/
+
+    override public T* pop ( )
+    {
+        if (this.buffer.length <= this.oldest)
+            return null;
+
+        return &this.buffer[][this.oldest++];
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of items in the queue
+
+    ***************************************************************************/
+
+    override public size_t length ( )
+    {
+        return this.buffer.length - this.oldest;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            number of bytes stored in queue
+
+    ***************************************************************************/
+
+    override public ulong used_space ( )
+    {
+        return this.length * T.sizeof;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            number of bytes free in queue
+
+    ***************************************************************************/
+
+    override public ulong free_space ( )
+    {
+        return this.oldest * T.sizeof;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            total number of bytes used by queue (used space + free space)
+
+    ***************************************************************************/
+
+    override public ulong total_space ( )
+    {
+        return this.buffer.length * T.sizeof;
+    }
```

Comment:
Do you think there'd be any reason to add a method to query the capacity of the buffer?

Reply:
`Buffer` struct currently does not expose it and in general it is beyond our control (runtime-defined). I can add it if you can think of at least one use case :)

Reply:
Ah ok. No, not really.